### ISSUE TITLE
TF-Slim hangs in sess.run()

### ISSUE THREAD

**Issue by i-akbari** [tensorflow/models#1162]
I want to run this example of Using TF-Slim:
http://warmspringwinds.github.io/tensorflow/tf-slim/2016/10/30/image-classification-and-segmentation-using-tensorflow-and-tf-slim/
but the programs hangs in this line. (I added two print defore and after)
```
print('before...')
np_image, network_input, probabilities = sess.run([image, processed_image, probabilities])
print('after')
```
**the problem is "after" will never displayed.**

The weblog post has used tensorflow-0.11 but I have used tensorflow-1.0 and tensorflow-1.0.1. I also upgraded tf-slim with tf_upgrade.py to be compatible with new tensorflow. After that, I installed tensorflow-0.11 but result was the same.

_Output on GPU:_
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
> I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:910] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
> name: GeForce GTX TITAN
> major: 3 minor: 5 memoryClockRate (GHz) 0.8755
> pciBusID 0000:01:00.0
> Total memory: 5.94GiB
> Free memory: 5.76GiB
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y 
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN, pci bus id: 0000:01:00.0)
> before...
> 
and the programs hangs here.
**What is the problem?**

_Output on CPU only:_
I switch to CPU mode with line below:
`os.environ["CUDA_VISIBLE_DEVICES"] = ''`

> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
> I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
> E tensorflow/stream_executor/cuda/cuda_driver.cc:509] failed call to cuInit: CUDA_ERROR_NO_DEVICE
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:158] retrieving CUDA diagnostic information for host: akbari
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:165] hostname: akbari
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:189] libcuda reported version is: 375.39.0
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:363] driver version file contents: """NVRM version: NVIDIA UNIX x86_64 Kernel Module  375.39  Tue Jan 31 20:47:00 PST 2017
> GCC version:  gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04.3) 
> """
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:193] kernel reported version is: 375.39.0
> I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:300] kernel version seems to match DSO: 375.39.0
> before...

The same problem.

**Code**
The whole program is:
```
from sqlalchemy.orm.scoping import scoped_session
from matplotlib import pyplot as plt
import numpy as np
import os
import tensorflow as tf
import urllib2
import sys
from datasets import imagenet
from nets import inception_v4, inception_resnet_v2, vgg
from preprocessing import vgg_preprocessing, inception_preprocessing

checkpoints_dir = '.../checkpoints/'

slim = tf.contrib.slim
os.environ["CUDA_VISIBLE_DEVICES"] = '0'

image_size = inception_v4.inception_v4.default_image_size

with tf.Graph().as_default():
    filenames = ['.../First_Student_IC_school_bus_202076.jpg']
    filename_queue = tf.train.string_input_producer(filenames)

    reader = tf.Whofrom sqlalchemy.orm.scoping import scoped_session
from matplotlib import pyplot as plt
import numpy as np
import os
import tensorflow as tf
import urllib2
import sys
from datasets import imagenet
from nets import inception_v4, inception_resnet_v2, vgg
from preprocessing import vgg_preprocessing, inception_preprocessing

checkpoints_dir = '.../checkpoints/'

slim = tf.contrib.slim
os.environ["CUDA_VISIBLE_DEVICES"] = '0'

image_size = inception_v4.inception_v4.default_image_size

with tf.Graph().as_default():
    filenames = ['.../First_Student_IC_school_bus_202076.jpg']
    filename_queue = tf.train.string_input_producer(filenames)

    reader = tf.WholeFileReader()
    key, value = reader.read(filename_queue)

    # image = tf.image.decode_jpeg(value)
    image = tf.image.decode_jpeg(value, channels=3)

    processed_image = inception_preprocessing.preprocess_image(image,
                                                           image_size,
                                                           image_size,
                                                           is_training=False)


    processed_images = tf.expand_dims(processed_image, 0)

    with slim.arg_scope(inception_v4.inception_v4_arg_scope()):
        logits, _ = inception_v4.inception_v4(processed_images,
                           num_classes=1001,
                           is_training=False)
    probabilities = tf.nn.softmax(logits)

    init_fn = slim.assign_from_checkpoint_fn(
    os.path.join(checkpoints_dir, 'inception_v4.ckpt'),
    slim.get_model_variables('InceptionV4'))

    with tf.Session() as sess:
        init_fn(sess)

        print('before...')
        np_image, network_input, probabilities = sess.run([image, processed_image, probabilities])
        # np_image, network_input, probabilities = sess.run([processed_image, probabilities])
        print('after!')

        probabilities = probabilities[0, 0:]
        sorted_inds = [i[0] for i in sorted(enumerate(-probabilities),
                                        key=lambda x: x[1])]

        # Show the downloaded image
        plt.figure()
        plt.imshow(np_image.astype(np.uint8))
        plt.suptitle("Downloaded image", fontsize=14, fontweight='bold')
        plt.axis('off')
        plt.show()

        plt.imshow(network_input / (network_input.max() - network_input.min()))
        plt.suptitle("Resized, Cropped and Mean-Centered input to network",
                 fontsize=14, fontweight='bold')
        plt.axis('off')
        plt.show()

        names = imagenet.create_readable_names_for_imagenet_labels()
        for i in range(5):
            index = sorted_inds[i]

            print('Probability %0.2f => [%s]' % (probabilities[index], names[index + 1]))
        res = slim.get_model_variables()
print('End!')
```

**TF-Slim version and upgrade:**
Because tf-slim is not compatible with Tensorflow 1.0 I used tf_upgrade.py script in https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/compatibility
Note that tf-slim is campatible with tensorflow-0.11 and there is no need to upgrade it.
_steps:_
1. git clone https://github.com/tensorflow/models
2. mv models models-old
3. tf_upgrade.py --intree models-old --outtree models
4. I modified this two problem manually:

    detected 2 errors that require attention
    models-old/street/python/nn_ops.py:186: tf.reverse requires manual check.
    models-old/street/python/nn_ops.py:249: tf.reverse requires manual check.

    modifications:
    line 186:
    #inp = tf.reverse(inp, [False, True, False])]
    inp = tf.reverse(inp, [1])

    line 249:
    #out = tf.reverse(out, [False, True, False])
    out = tf.reverse(out, [1])

5. I added models/slim to python path.

**OS, Python, Tensorflow, Cuda, cudnn, Driver and GPU**

_with all these version the problem exist:_
> OS:
> Ubuntu 14.04.5 LTS
> Debian GNU/Linux 8 (jessie)
> 
> Python-2.7.6
> 
> tensorflow-0.11
> tensorflow-1
> tensorflow-1.0.1
> 
> Cuda-8.0.27
> Cuda-8.0.61
> 
> cudnn-4
> cudnn-5
> cudnn-5.1
> 
> NVIDIA driver:
> NVIDIA-Linux-x86_64-361.??
> NVIDIA-Linux-x86_64-367.27
> NVIDIA-Linux-x86_64-367.44
> NVIDIA-Linux-x86_64-375.39
> 
> GPU:
> GeForce GTX TITAN
> Tesla K20m

**Comment by concretevitamin** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by i-akbari** [comment]
I think it is a bug!

**Comment by concretevitamin** [comment]
@i-akbari: we'd be more than happy to look into more details if you could provide a minimum reproducible program.

In this case, now that I looked a bit more, could it be that you've failed to call `start_queue_runners()`?  See [here](https://www.tensorflow.org/programmers_guide/reading_data#creating_threads_to_prefetch_using_queuerunner_objects).

**Comment by i-akbari** [comment]
@concretevitamin Thank you for your response. I don't know that these information are useful or not.
I flowed function calls and find out program hangs on return line.
```
def _run_fn(session, feed_dict, fetch_list, target_list, options,
                run_metadata):
      # Ensure any changes to the graph are reflected in the runtime.
      self._extend_graph()
      with errors.raise_exception_on_not_ok_status() as status:
        return tf_session.TF_Run(session, options,
                                 feed_dict, fetch_list, target_list,
                                 status, run_metadata)
```

The debuger shows these values: 

> session: Swig Object of type 'TF_DeprecatedSession *' at 0x7f3b40185ba0
> options: None  
> feed_dict: {}  
> fetch_list: type 'list': ['DecodeJpeg:0', 'eval_image/Mul:0', 'Softmax:0']  
> target_list: type 'list': []  
> status: Swig Object of type 'TF_Status *' at 0x7f3b40185c00  
> run_metadata: tensorflow.python.pywrap_tensorflow.TF_Buffer; proxy of Swig Object of type 'TF_Buffer *' at 0x7f3b40185e70

**Comment by concretevitamin** [comment]
@i-akbari are you sure it's not something as simple as not starting queue runners?

**Comment by i-akbari** [comment]
@concretevitamin Where exactly should I start queue runners? Could you give me the code?

**Comment by concretevitamin** [comment]
Could you take a look at the link I pasted in a previous comment?

**Comment by i-akbari** [comment]
Thank you for your response. My code is a full example so there is no need to change it. Any way I did my work with Keras. I couldn't solve the problem.
